### PR TITLE
Refs #29983 -- Added pathlib.Path support to the file email backend.

### PR DESCRIPTION
**Pull request by jdufresne**

### CODE REVIEW

File: django/core/mail/backends/filebased.py [hunk 1]
```diff
@@ -17,9 +17,6 @@ def __init__(self, *args, file_path=None, **kwargs):
             self.file_path = file_path
         else:
             self.file_path = getattr(settings, 'EMAIL_FILE_PATH', None)
-        # Make sure self.file_path is a string.
-        if not isinstance(self.file_path, str):
```
Comment (pope1ni) [inline django/core/mail/backends/filebased.py]: Reinstate this, but add `pathlib.Path` to the `isinstance` check. The comment was staying the obvious and can go.

File: docs/releases/3.1.txt [hunk 2]
```diff
@@ -119,7 +119,8 @@ CSRF
 Email
 ~~~~~
 
-* ...
+* The :ref:`file email backend <topic-email-file-backend>` now supports
```
Comment (pope1ni) [inline docs/releases/3.1.txt]: Explicitly mention the setting here. It's not really clear what is meant by "supports".

File: django/core/mail/backends/filebased.py [hunk 3]
```diff
@@ -17,9 +17,6 @@ def __init__(self, *args, file_path=None, **kwargs):
...
-            raise ImproperlyConfigured('Path for saving emails is invalid: %r' % self.file_path)
```
Comment (pope1ni) [inline django/core/mail/backends/filebased.py]: With this reinstated, it looks as though a test is needed as you didn't remove one...

File: docs/releases/3.1.txt (hunk 2 above)
Comment (jdufresne) [reply docs/releases/3.1.txt]: Done. I'm open to changing the wording if there are more suggestions.

File: django/core/mail/backends/filebased.py (hunk 1 above)
Comment (jdufresne) [reply django/core/mail/backends/filebased.py]: I had considered this direction on the first approach but chose to remove the type checking instead. I feel that this better follows Python's duck typing tradition. The value of `self.file_path` is always passed to `os.path.abspath()` anyway which does its own type checking, so Django doesn't need to duplicate it. So invalid values are still rejected.

I've added a test for the `os.path.abspath()` type checking.

**Comment by pope1ni** [comment]
Oops. I think we missed an actual [test failure](https://djangoci.com/job/pull-requests-windows/database=sqlite3,label=windows,python=Python38/2661/testReport/junit/mail.tests/MailTests/test_backend_arg/) on this:

```
TypeError: _getfullpathname: path should be string, bytes or os.PathLike, not object

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Jenkins\workspace\pull-requests-windows\database\sqlite3\label\windows\python\Python38\tests\mail\tests.py", line 574, in test_backend_arg
    mail.get_connection('django.core.mail.backends.filebased.EmailBackend', file_path=object())
  File "C:\Python38\lib\contextlib.py", line 131, in __exit__
    self.gen.throw(type, value, traceback)
  File "C:\Jenkins\workspace\pull-requests-windows\database\sqlite3\label\windows\python\Python38\django\test\testcases.py", line 683, in _assert_raises_or_warns_cm
    self.assertIn(expected_message, str(getattr(cm, cm_attr)))
AssertionError: 'expected str, bytes or os.PathLike object, not object' not found in '_getfullpathname: path should be string, bytes or os.PathLike, not object'
```

Perhaps we should reconsider https://github.com/django/django/pull/12022#discussion_r342896572?

Strange that it only occurred for the Windows builder.

**Comment by carltongibson** [comment]
Ah, well done Nick. I shall have a look this afternoon.

**Comment by felixxm** [comment]
https://github.com/python/cpython/blob/353fb1ecbfd58752dabae115c4964095e1e35e5f/Lib/ntpath.py#L508-L521

**Comment by felixxm** [comment]
`nt._getfullpathname` is used only on Windows.

**Comment by carltongibson** [comment]
Good work @felixxm! I'll do a PR shortly. 

(Grrr. Can't believe the Windows failure was actually a failure! 😡)

**Comment by jdufresne** [comment]
I opened #12032 to fix the failure.

**Comment by pope1ni** [comment]
> Can't believe the Windows failure was actually a failure! 😡

The one time you assume...

**Comment by carltongibson** [comment]
> Perhaps we should reconsider #12022 (comment)?

Just to address your comment @pope1ni, I was happy with Jon's line that the test exercises that the type error is caught, so we can reasonably _duck-type_ here, so that's why I left it as was.

**Pull request merged**